### ISSUE TITLE
Replication problems

### ISSUE THREAD

**Issue by pfrazee** [datproject/dat#635]
Hate to say it but we've got still got replication issues lurking.

I ran this command, to fetch the datprotocol site:

```
dat dat://ff34725120b2f3c5bd5028e4f61d14a45a22af48a7b12126d5d588becde88a93/ ~/tmp/datprotocol.com
```

I noticed it would make small bits of progress, then halt, then progress, then halt. Good sign that the connection is getting killed. So I ran with DEBUG=* and [this is the output](https://gist.github.com/pfrazee/3416654095f3080e13113c3cea03afd1). There's at least one checksum error, but most of the time either the connection is silently being killed, or it's being killed from the other side.

**Comment by pfrazee** [comment]
$10 says this is still related to missing historical data.

**Comment by joehand** [comment]
> $10 says this is still related to missing historical data.
 
Can I pay in frazeebucks instead?

Definitely seems like a more recent regression. I'll make a branch that stores historic data and we can test from there.

**Comment by pfrazee** [comment]
Spent some time watching logs on my local client and on my cloud server, both running latest dat from the CLI. I am able to reproduce:

First I should mention: The cloud server is only able to download 92% of the dat from my original source, which is my beaker instance. I cleared data and tested this again to confirm. I'm pretty sure this is reflective of the fact that history isn't being kept (8% of it isnt kept, in fact).

Then I close my original source, beaker, and download from the dat CLI. There are no errors thrown in the cloud server log in my runs this evening, but plenty of errors in my local dat process: failed signature checks, failed checksums, etc.

Speculating: Is it possible this is a regression that has to do with @mafintosh recent update to hyperdrive, to keep a capture in memory?

Another question: is dat doing verify-on-read? If it is, then it shouldn't be able to send bad data, because it should be verifyhing the checksum before sending.

> frazeebucks

Nice try! You may think you know the worth of a frazeebuck, but to me, frazeebucks are worthless

**Comment by pfrazee** [comment]
@mafintosh I discovered there is another peer involved, so when I debugged Beaker->My_Cloud, as discussed, I didn't get new info.

New idea. I'm looking at your [selectLatest](https://github.com/mafintosh/hyperdrive/blob/master/archive.js#L47) code and wondering if it's correct. While metadata is still being downloaded, isn't the concept of "latest" changing rapidly? So, any request to download latest will probably need to be canceled -- or at least processed *in order*. Is it possible there's a race here?

**Comment by joehand** [comment]
Still seeing some issues with local CLI usage too, always replicating from the source:

It seems like whenever a client misses an update then takes two updates to get the content back to the latest, a few situations like that:

#### Misses Initial History

A client misses some initial history before the first clone:

Source:
* Create dir: `mkdir toshare && echo "hello" > toshare/hello.txt"`
* Add files: `dat share toshare` (keep this going throughout test)
* (separate process): `echo "more" >> toshare/hello.txt`

Client:
* (after last bullet above) initial clone: `dat clone <link> isclient`
* `cat isclient/hello.txt` prints `hello`

If you update the source again, the client will catch up:

* (source) `echo "third update" >> toshare/hello.txt`
* (client) `dat pull isclient`
* (client) `cat isclient/hello.txt` prints `hello\n more\n thirdupdate`

During this process `archive.metadata` feed is always correct and `archive.content.blocks` is also correct.

**Comment by joehand** [comment]
Reproducible test case:

First, start sharing an empty dir:
```
mkdir toshare
dat share toshare
```

Add file to dir and update a few times (with live watch turned on this should be updating metadata):
```
echo "hello" >> toshare/hello.txt
echo "something with lots of bytes" >> toshare/hello.txt
```

Then clone somewhere:
```
❯ dat clone <link> isclient
❯ cat hello.txt 
hello
```

`hello.txt` does not get the second update. The metadata seems to get updated (the size is correct) and `archive.content.blocks` is also correct.

**Comment by joehand** [comment]
The `dat-node` patch may have fixed above test but got another strange test case. In this case, you can get the correct content from a client (through another clone) but the file does not show correct content.


Follow similar test case above (it seems like the first clone is okay now, but further ones fail):

---

First, start sharing an empty dir:
```
mkdir toshare
❯ dat share toshare
```

Add file to dir and update a few times (with live watch turned on this should be updating metadata):
```
❯ echo "1 line" >> toshare/hello.txt
❯ echo "2 line" >> toshare/hello.txt
```

Then clone somewhere:
```
❯ dat clone <link> client-one
❯ cat client-one/hello.txt 
1 line
2 line
```

Add another line to source file (and make sure it is correct still)
```
❯ echo "3 line" >> toshare/hello.txt
❯ cat toshare/hello.txt
1 line
2 line
3 line
```

Update the initial clone with a pull. The file should still have 2 lines (incorrectly)
```
❯ dat pull client-one
❯ cat client-one/hello.txt
1 line
2 line
```

**Stop the source `dat share` process**

Sync data from the first client:
```
❯ dat sync client-one 
```

Clone to another client (this should be pulling from `client-one`)
```
❯ dat clone <link> client-two
❯ cat client-two/hello.txt
1 line
2 line
3 line
```

The file should have ALL the lines even though the client it pulled from only has 2 lines.

**Comment by joehand** [comment]
A few observations from debugging this morning.

These bugs are definitely related to this PR https://github.com/mafintosh/hyperdrive/pull/123. If I use hyperdrive `7.12.2` then everything works fine (without taking into account clients that have content history stored).

On the sharing side:

Setting `opts.file` to
```js
opts.file = function (name, opts) {
  return raf(path.join(dir, name), opts && typeof opts.length === 'number' && {length: opts.length})
}
```

Breaks `archive.createFileReadStream()`, which is part of why I think its overwriting the source sometimes. Still trying to see if that is true on download side too.

**Comment by yoshuawuyts** [comment]
@joehand if `hyperdrive@7.12.2` is stable, is there a chance we can pin it to that for the mean time? This could perhaps with our upstream issue of https://github.com/datproject/dat-desktop/issues/210, at least so users aren't affected.

**Comment by joehand** [comment]
Released dat-node 1.3.6 with that pinned and a few other bug fixes. Still seeing replication issues and trying to track those down. 

Related:

* dat-node PR for release: https://github.com/datproject/dat-node/pull/100
* hypercore replication bug: https://github.com/mafintosh/hypercore/issues/69#issuecomment-279813844
* hyperdrive truncation bug PR: https://github.com/mafintosh/hyperdrive/pull/125

**Comment by joehand** [comment]
Fixed? :crossed_fingers:
